### DISCUSSION TITLE
Software calibration , solve mechanical problem in CNC

### DISCUSSION THREAD

**Discussion by dzario** [Paciente8159/uCNC#272] [Ideas]
Hi. 
I have idea add functions to solve mechanical problem in CNC (right angle, axe deviation .. etc .).
Using calibrated rod , measure errors and then save to firmware, function use this data and in motion 
adapt .

Paciente8159 give my your email pls .

regards.

**Comment by Paciente8159** [comment]
Hi there. 
Feel free to share any ideas in this discussion.
I can then check if the current features of µCNC already support what you need or not.
Sorry no private messages. 

Thanks,

**Reply by dzario** [reply]
Hi. OK.
Ehm it's will be long.. 

I'm hobbyist and in my short life build many CNC machine, it's my pleasure from nothing make machine what can anything .
Look to my photo album : [CNC](https://photos.app.goo.gl/n6jVv6HNPqZgrybJ6) , [CNC1](https://goo.gl/photos/GG2egEVPpwm9rsH49) ,
[CNC2](https://goo.gl/photos/dT1RxMb4zba1Hh6k6) and here [CNC3](https://goo.gl/photos/dT1RxMb4zba1Hh6k6) .
My last CNC is made from polymer concrete ( epoxide , sand and rock ) .
Very heavy part .. .
There is not problem to build CNC machine, very big problem is build accurate machine.

Then idea is make software compensation for all axes .
Look at this photo : [X axis tune](https://photos.app.goo.gl/uNrsxJbstFYxVJeA7)
I use thin aluminium strip and try tune mechanical problem , I want accuracy better than 0.01mm . But if I do it , then still have problem with other direction. All hobbyist buy linear rail form China, me too , and this is not bes quality. Rod is bend, twisted etc .. and my mold make CNC part also not perfect . Then machine is stable, heavy , super for resonance , very strong .... but not accurate.

1. Make function in firmware to measure surface of very accurate rod (I got it , very accurate with deviation +- 0.007mm) and every 1cm for touch probe measure . For example in X axes is rod and X move every 1cm and touch surface , Z value store to table . In length 1m there will be 100 measure point, if use 2bytes then use 200bytes memory . This data then store to firmware. Later if X axes will move then read from this table and compensate Z axes . 

2. It's similar to mill PCB . Now I use EstlCam software , this soft have surface scan to mill PCB or mill curved surface
With this is possible mill exactly value from surface.

3. Not only linearity , also right angle between axes X vs Y , X vs Z ..etc .. and skew .. .. .

I don't want go to professional CNC company and mill or grind surface to screw linear rail , this not help , because used linear rail is not best quality , I see it in my machine. 

Write to me , is possible add to your firmware this function, may be I can help to you and donate some money for your time .
This last CNC machine is my dead :( , I try solve mechanical problem lots of days and still not good . If have software compensation then all will be more easy .

Best regards.

**Comment by Paciente8159** [comment]
I like your CNC machines.

Ok. I think I understood your problem and I'll try to help the best way I can.

### Surface correction
Even if you have a very flat surface, bended linear guides (apart from some possible problems in the motion of the machine) will make your machine (gantry, tool or bed) move in an non linear way. Like this image:

In that case I think the best possible solution is to use a surface mapping that modifies the Gcode on the fly to compensate for that.
![Surface Map](https://forum.linuxcnc.org/media/kunena/attachments/24854/compensationMap.png)
In the present moment µCNC by itself does not support this feature. I want to implement this feature but I have other features that I want to implement before picking this up. There are software that do this task that can be used with µCNC or Grbl to create a surface map and improve this problem. Please check [OpenCNCPilot](https://github.com/martin2250/OpenCNCPilot) or [bCNC](https://github.com/vlachoudis/bCNC). They can create a map of the surface that you apply to your Gcode to correct the surface problem.
OpenCNCPilot for example will enable you to save the map file and reapply it every time you need without having to redo the probing. bCNC probably allows the same but I'm not sure.

If you need to do it in firmware then [Marlin](https://github.com/MarlinFirmware/Marlin) or [Klipper](https://www.klipper3d.org/) might be an alternative.

I recommend getting a Renishaw probe or similar (or building one) to do the mapping surface. 

About the suggestion of using an accurate rod to measure any distortion of the axis there are 2 problems with that.

- First you are only measuring distortion in one axis. If you have a distortion in any other axis it would not solve.
- The other is the amount of data is bigger than that. You need 4bytes (float) per point and a bi-dimensional array and not a single dimension array. On a grid of 100x100 that is a bit more than 39Kbytes in data. Usually lesser points are taken, and then some sort of bilinear surface interpolation is used.

### Axis non perpendicular (skew)
Yes µCNC as that implemented and can be enabled. It works like skew in Marlin. Once enable 2 additional settings will be available to enter the correction factors. The factors are calculated like in Malin. Here is the explanation from Marlin firmware:

```
/**
 * Bed Skew Compensation
 *
 * This feature corrects for misalignment in the XYZ axes.
 *
 * Take the following steps to get the bed skew in the XY plane:
 *  1. Print a test square (e.g., https://www.thingiverse.com/thing:2563185)
 *  2. For XY_DIAG_AC measure the diagonal A to C
 *  3. For XY_DIAG_BD measure the diagonal B to D
 *  4. For XY_SIDE_AD measure the edge A to D
 *
 * Marlin automatically computes skew factors from these measurements.
 * Skew factors may also be computed and set manually:
 *
 *  - Compute AB     : SQRT(2*AC*AC+2*BD*BD-4*AD*AD)/2
 *  - XY_SKEW_FACTOR : TAN(PI/2-ACOS((AC*AC-AB*AB-AD*AD)/(2*AB*AD)))
 *
 * If desired, follow the same procedure for XZ and YZ.
 * Use these diagrams for reference:
 *
 *    Y                     Z                     Z
 *    ^     B-------C       ^     B-------C       ^     B-------C
 *    |    /       /        |    /       /        |    /       /
 *    |   /       /         |   /       /         |   /       /
 *    |  A-------D          |  A-------D          |  A-------D
 *    +-------------->X     +-------------->X     +-------------->Y
 *     XY_SKEW_FACTOR        XZ_SKEW_FACTOR        YZ_SKEW_FACTOR
 */
```

I'm going to be honest. A precision of 0.01mm or lower is not easy on a hobby machine. If you need other or more complex software corrections than you might want to consider using [LinuxCNC](https://linuxcnc.org/).

One final note. You don't have to donate anything. I'll gladly help you in any way I can.
Trying µCNC and report any issue that you may find is also very helpful.

Hope I've covered your questions.

Regards,

**Reply by dzario** [reply]
Hi.
Ehm .. no , My idea is different..
I don't want measure surface , I don't need it . 
Surface scan adapt gcode to surface , perfect for 3D print, engrave PCB etc .. or mill to bending part .

If finish CNC then last operation is mill surface . 
My project right now , I do it , but if not solve mechanical problem , then surface copy X and Y axis deviation.

We don't need 2 dimensional matrix with touch point, and 4 bytes.
Just 2 bytes, for X axes one dimension, first point is 0 (doesn't mater what Z is value) , all next point are + or - from starting point, store step , not real Z axis position.
I try explanation :

1. use calibrated accurate rod , and stand on two point .
![calibrod](https://user-images.githubusercontent.com/47868181/192142275-ee02169d-6a65-4feb-8a34-4a325ecb2562.png)
and use shim rings or thin metal strip to set : left side 0  (dial indicator), and right side also 0 (dial indicator) .
Then deviation are between two point , min left , and max right of X axis .
In my case it's maximum -0.05mm . Then if my motor use 800step per mm , 800/0.05 = 40 step is max . it's easy integer number, 
2. then we can start process to measure every 10mm with touch probe or every 5mm or .. less .. and store only differences .
3. after this I can mill working place and we will have accurate surface in X axes .

Same do in Y axes . Software calibration can't solve big mechanical mistake in short distance, 
If made machine with X axes 100mm travel distance and have 2mm +- deviation , then my machine is totally scrap.
But if make machine and in 1000mm have deviation 2mm , it's OK .

X axes is ideal for starting point , because in real world distance between linear rail is around 150mm or max 250mm . 
For example in my last machine is 180mm , in Datron ML1500 around 200mm etc .. 
In Y axes there are big distance .
Z axes is in real world short , with no long travel distance. In my practice Z axes move no more then 50mm , then machine error not big. 

Best regards.

PS: I have one black pill and test it .. .

**Reply by dzario** [reply]
Ehm ... just one idea ..
this function sove my problem .. I tested in Excel .. : 
z = z+(9 * 10^-11 * x^4)-(6 * 10^-8 * x^3)+(9 * 10^-6 * x^2)+(8 * 10^-5 * x)-0,01

this function generate Excel .. and perfect fit to my X axes .. 

regards.

**Comment by Paciente8159** [comment]
That might work but only partially. I'll try to explain my thoughts on the matter.

If I'm understanding right you are building a moving gantry type of machine (like your CNC and CNC1). That is a fixed bed and a gantry that moves the in the Y direction. The gantry carries the tool and moves it X and Z.

That plan of yours might work for X (were the tool moves right and left) because the linear guides are close to each other. But for the Y it may not. Unless you have exactly the same bending on both Y linear guides of your machine the gantry will not move in the Y axis in an uniform way.
I'll use the same image above to explain. 

![plane](https://user-images.githubusercontent.com/12116536/192163997-a93caa8a-2523-496b-a1bc-8301f513f496.png)

If the bending happens in different places the deformation will not be a wave but a concave/convex plane.

I have 2 suggestions for you. Either get a super flat surface and do the surface scanning there.

Or use your method on both sides of Y axis.
Then try to use a tool can convert your measurements to a plane (Matlab or something of that sort) like the one I used in the example. From there you can export your result to excel.

In excel you should be able to generate an hmap for OpenCNCPilot ( that is nothing more than a XML file with the mapped points) and see if this fixes the motion for you. Once done apply the same hmap to all your running Gcode.

Remember your origin reference must always be the same. I you move your homing point from there you have to repeat the process.

Have you tried help in other forums? I think that I've seen some concrete epoxy CNC builds over at [www.cnczone.com](www.cnczone.com)

Hope it helps.

**Comment by Paciente8159** [comment]
Just one more note. I think you should probably address the skew issue before mapping.
First make it move like a cube. Then take care of the uneven surface with the mapping.

**Comment by dzario** [comment]
Hi.
I litle test OpenCNCPilot . Interesting software. Test with Probing (soft only , not in real machine) working .
Any software what can solve skew issue? Gcode transformation ... any idea ?

Best regards.

**Reply by Paciente8159** [reply]
Sure. µCNC should be able to do this in the firmware.
There are also software options. There was GSuite in the cncforum but it does not seem to be available. I did not try but you can try [python script](https://github.com/MechanizedMedic/gskewer)

**Reply by dzario** [reply]
Hi , thanks for link .. I try it ..
I find this on youtube :  [Error Compensation of CNC Machines - Part 1/4](https://www.youtube.com/watch?v=pyj93ZXWUB8&ab_channel=InternationalPrecisionEngineering) , but link with code not workink.. I try script but not today ..

Best regards.

**Reply by dzario** [reply]
I test script :  
`E:\Dominik\Gcode>skew.py --xyerr 1.2 --xylen 80 test_stvorec_100_100.nc
usage: skew.py [-h] (--xyerr XYERR | --xytan XYTAN) (--yzerr YZERR | --yztan YZTAN) (--zxerr ZXERR | --zxtan ZXTAN)
               [--xylen XYLEN] [--yzlen YZLEN] [--zxlen ZXLEN]
               file
skew.py: error: one of the arguments --yzerr --yztan is required`

not workinig , required paramater , but I give him all parameter .. what is wrong ..?
imput ifle is simple sqare 100mm x 100mm ,

**Reply by Paciente8159** [reply]
As I told you I did not try it.. and yes it seems it is not working. Simply deletes my gcode and does nothing.
Found [GSuite](https://mega.nz/file/Y6gTUAwC#XQgcda_r7zaA9C3QaKFtNdDzRdGr9GlWxlGnmnlHA2c). Try this one.

**Reply by dzario** [reply]
yes , I laso find GSuite .. now try it .. 

best regards.

**Reply by Paciente8159** [reply]
Beware that GSkew only deals with X and Y skew.
More than you will need to find another application or try the firmware skew settings.
